processes/bootstrap/plugin.go

```diff
@@ -0,0 +1,224 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bootstrap
+
+import (
+	"github.com/ligato/cn-infra/infra"
+	pm "github.com/ligato/cn-infra/processes/processmanager"
+	"github.com/ligato/cn-infra/processes/processmanager/status"
+	"github.com/pkg/errors"
+	"os"
+	"sync"
+)
+
+// Bootstrap defines methods to obtain information about running instances
+type Bootstrap interface {
+	// GetProcessNames returns names of all running process instances
+	GetProcessNames() []string
+	// GetProcessByName returns an instance of given process
+	GetProcessByName(name string) pm.ProcessInstance
+}
+
+// Plugin is a bootstrap plugin structure
+type Plugin struct {
+	mx sync.Mutex
+
+	// a list of active process instances
+	instances map[string]pm.ProcessInstance
+
+	config *Config
+	Deps
+}
+
+// Deps are bootstrap dependencies (standard infra plugins + process manager)
+type Deps struct {
+	pm pm.ProcessManager
+	infra.PluginDeps
+}
+
+// Init starts processes and their watchers defined by the configuration file
+func (p *Plugin) Init() error {
+	p.instances = make(map[string]pm.ProcessInstance)
+
+	if err := p.getConfig(); err != nil {
+		return err
+	}
+
+	p.start(p.config.Processes)
+	p.watch()
+
+	return nil
+}
+
+// Close stops all remaining processes
+func (p *Plugin) Close() error {
+	p.mx.Lock()
+	defer p.mx.Unlock()
+
+	// TODO option to keep them running?
+
+	for name, instance := range p.instances {
+		if _, err := instance.StopAndWait(); err != nil {
+			p.Log.Errorf("bootstrap close error: failed to stop process %s: %v", name, err)
+		}
+	}
+
+	return nil
+}
+
+// String is a plugin name
+func (p *Plugin) String() string {
+	return p.PluginName.String()
+}
+
+// GetProcessNames returns names of all running process instances
+func (p *Plugin) GetProcessNames() (names []string) {
+	for name := range p.instances {
+		names = append(names, name)
+	}
+	return names
+}
+
+// GetProcessByName returns an instance of given process
+func (p *Plugin) GetProcessByName(reqName string) pm.ProcessInstance {
+	for name, instance := range p.instances {
+		if name == reqName {
+			return instance
+		}
+	}
+	return nil
+}
+
+func (p *Plugin) start(processes []Process) {
+	for _, process := range processes {
+		if err := p.validate(&process); err != nil {
+			p.Log.Errorf("unable to start process %s: %v", process.Name, err)
+			continue
+		}
+		processStat := make(chan status.ProcessStatus)
+		pLogger := newPmLogger(process.Name, process.LogFilePath)
+		instance := p.pm.NewProcess(process.Name, process.BinaryPath, pm.Args(process.Args...),
+			pm.Writer(pLogger, pLogger), pm.Notify(processStat), pm.AutoTerminate())
+		if err := instance.Start(); err != nil {
+			p.Log.Errorf("Error starting process %s: %v", process.Name, err)
+			os.Exit(-1)
```

Comment:
Plugins should not exit process. The `start` function should return error and let caller handle it.

Reply:
Good point